homeassistant/components/tesla_fleet/oauth.py

```diff
@@ -0,0 +1,77 @@
+"""Application Credentials platform the Tesla Fleet integration."""
+
+import base64
+import hashlib
+import secrets
+from typing import Any
+
+from homeassistant.components.application_credentials import (
+    AuthImplementation,
+    AuthorizationServer,
+    ClientCredential,
+)
+from homeassistant.core import HomeAssistant
+
+from .const import AUTHORIZE_URL, CLIENT_ID, DOMAIN, SCOPES, TOKEN_URL
+
+AUTH_SERVER = AuthorizationServer(AUTHORIZE_URL, TOKEN_URL)
+
+
+class TeslaSystemImplementation(AuthImplementation):
+    """Tesla Fleet API open source Oauth2 implementation."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize open source Oauth2 implementation."""
+
+        # Setup PKCE
+        self.code_verifier = secrets.token_urlsafe(32)
+        hashed_verifier = hashlib.sha256(self.code_verifier.encode()).digest()
+        self.code_challenge = (
+            base64.urlsafe_b64encode(hashed_verifier).decode().replace("=", "")
+        )
+        super().__init__(
+            hass,
+            DOMAIN,
+            ClientCredential(CLIENT_ID, "", "Built-in open source client ID"),
```

Comment:
Don't use anything from application credentials integration in this implementation.